### PR TITLE
[SCB-2077] fix instance isolation trying chance not released problem

### PR DESCRIPTION
**Pull request by yhs0092**
Please see the details in https://issues.apache.org/jira/browse/SCB-2077

### CODE REVIEW

**Comment by yhs0092** [comment]
Current fix solution:
- change the type of `ServiceCombServerStats#globalAllowIsolatedServerTryingFlag` from `AtomicBoolean` to `AtomicReference<TryingIsolatedServerMarker>`. The newly defined `TryingIsolatedServerMarker` holds the `Invocation` instance which is holding the trying chance and the timestamp.
- in the normal case, when the `Invocation` is finished(checked by `Invocation#isFinished`), the trying chance is released.
- in case that some unknown error occurs, causing the invocation procedure ends with the `Invocation#finished` flag not set to true, an new configuration `servicecomb.loadbalance.isolation.maxSingleTestWindow` is added to limit the max occupation period of the global trying chance.

**Comment by yhs0092** [comment]
I've also considered about other solutions.
An alternative choice is to use `SoftReference` to hold the `Invocation` instance in `TryingIsolatedServerMarker`. In case that the `Invocation#finished` is not set to true properly, the trying chance can be released when the `SoftReference` returns `null` (which means the `Invocation` instance is garbage collected --- the invocation is over).
The advantage of this solution is the new config item `servicecomb.loadbalance.isolation.maxSingleTestWindow` is not needed. While there is also the disadvantage. Since the result of `SoftReference#get` gets null after the garbage collection happens, this insurance policy depends on the timing of GC. The result seems uncontrollable.

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/ServiceCombServerStats.java [hunk 1]
```diff
@@ -35,15 +36,14 @@
 
   private static final Logger LOGGER = LoggerFactory.getLogger(ServiceCombServerStats.class);
 
+  private final Object lock = new Object();
```
Comment (liubao68) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/ServiceCombServerStats.java:39]: lock is never read

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/ServiceCombServerStats.java [hunk 2]
```diff
@@ -84,20 +84,35 @@ private void init(){
...
   }
 
   /**
    * Applying for a trying chance for the isolated server. There is only 1 trying chance globally concurrently.
    *
    * @return true if the chance is applied successfully, otherwise false
    */
-  public static boolean applyForTryingChance() {
-    return isolatedServerCanTry() && globalAllowIsolatedServerTryingFlag.compareAndSet(true, false);
+  public static boolean applyForTryingChance(Invocation invocation) {
```
Comment (liubao68) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/ServiceCombServerStats.java:99]: This method is not thread safe.

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/ServiceCombServerStats.java (hunk 1 above)
Comment (yhs0092) [reply handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/ServiceCombServerStats.java:39]: The `lock` is used in method `ensureWindow`. It does not get involved in this issue.

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/ServiceCombServerStats.java (hunk 2 above)
Comment (yhs0092) [reply handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/ServiceCombServerStats.java:99]: The thread safty is ensured by the CAS operation `AtomicReference#compareAndSet`

**Pull request merged**